File: Android App/InOut_Nodeapp/app/src/main/java/com/example/inout_nodeapp/APSettingActivity.java
package com.example.inout_nodeapp;

import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.google.android.material.snackbar.Snackbar;

import static com.example.inout_nodeapp.MyApplication.AppColor;
import static com.example.inout_nodeapp.MyApplication.SelectedAPIndex;
import static com.example.inout_nodeapp.MyApplication.TextColor;
import static com.example.inout_nodeapp.MyApplication.getStringCode;
import static java.lang.String.format;

public class APSettingActivity extends AppCompatActivity {
    View mainLayout;
    boolean mIsBound;
    ListView List_AP;
    TextView Text_AP;
    Button ReloadBtn;
    MyApplication APP = (MyApplication) getApplication();
    Toolbar toolbar;
    ImageButton Home;
    TextView Title;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_ap_setting);
        mainLayout = findViewById(R.id.apSettingparent);
        SetToolBar();
        List_AP = findViewById(R.id.List_AP);
        Text_AP = findViewById(R.id.Text_AP);
        List_AP.setOnItemClickListener(this::SelectListView);
        ReloadBtn = findViewById(R.id.reload);
        ReloadBtn.setText("RELOAD");
        ReloadBtn.setOnClickListener(v-> requestAdminAPI());
        ((TextView)findViewById(R.id.textView)).setText(getStringCode(10108));

        setColor();//색 설정 : 툴바, 재시도 버튼

        Intent intent = new Intent(this,MyService.class);
        mIsBound = bindService(intent,mConnection, Context.BIND_AUTO_CREATE);
    }

    void SetToolBar()
    {
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        Title = findViewById(R.id.toolbarText);
        Title.setText(getStringCode(9998,getString(R.string.app_name)));
        actionBar.setDisplayShowTitleEnabled(false);
        Home = findViewById(R.id.imageBtn);
        Home.setOnClickListener((v)->{
            Intent intent = new Intent(this,MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            intent.putExtra("isNEW",false);
            startActivity(intent);
        });
    }
    void setColor()
    {
        toolbar.setBackgroundColor(Color.parseColor(AppColor));
        Title.setBackgroundColor(Color.parseColor(AppColor));//색 설정
        Title.setTextColor(Color.parseColor(TextColor));
        Home.setBackgroundColor(Color.parseColor(AppColor));
        Home.setColorFilter(Color.parseColor(TextColor));
        ReloadBtn.setBackgroundColor(Color.parseColor(AppColor));
        ReloadBtn.setTextColor(Color.parseColor(TextColor));
    }
    @Override
    public void finish() {
        try {
            mServiceMessenger.send(Message.obtain(null,MyService.ACTIVITY_1));

            Intent intent = new Intent(this,MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            intent.putExtra("isNEW",false);
            startActivity(intent);
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }
    boolean isNeedIPSet = false;
    @Override
    protected void onPostResume() {
        super.onPostResume();
        if(isNeedIPSet) {
            isNeedIPSet = false;
            requestAdminAPI();
        }
    }

    void requestAdminAPI()
    {
        try {
            mServiceMessenger.send(Message.obtain(null,MyService.GETADMIN));
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }
    class ActivityHandler extends Handler {
        @Override
        public void handleMessage(@NonNull Message msg) {
            switch (msg.what)
            {
                default: super.handleMessage(msg); break;
            }
        }
    }
    Messenger mServiceMessenger = null;
    final Messenger mMessenger = new Messenger(new APSettingActivity.ActivityHandler());
    private ServiceConnection mConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mServiceMessenger = new Messenger(service);
            try{
                Message msg = Message.obtain(null,MyService.CONNECT);
                msg.replyTo = mMessenger;
                mServiceMessenger.send(msg);

                mServiceMessenger.send(Message.obtain(null,MyService.ACTIVITY_3));//Sets the "where to send the reply" of the Service : This Activity

                mServiceMessenger.send(Message.obtain(null,MyService.GETADMIN));

                msg = Message.obtain(null,
                        MyService.SEND_VALUE,123456789,0);
                mServiceMessenger.send(msg);
            }catch (RemoteException e){}
        }

        @Override
        public void onServiceDisconnected(ComponentName name) { mServiceMessenger = null; }
    };

    @Override
    protected void onNewIntent(Intent intent) {
        processIntent(intent);
        super.onNewIntent(intent);
    }
    private void processIntent(Intent intent) {
        if(intent != null)
        {
            String reply = intent.getStringExtra("reply");

            if(reply == null)return;
            if(reply.equals("MessageAlert"))
            {
                String Title = intent.getStringExtra("Title");
                String Message = intent.getStringExtra("Message");
                MessageAlert(Title,Message);
            }
            if(reply.equals("MessageSnack"))
            {
                String Title = intent.getStringExtra("Title");
                String Message = intent.getStringExtra("Message");
                MessageSnack(Title,Message);
            }
            if(reply.equals("MessageTerminate"))
            {
                String Title = intent.getStringExtra("Title");
                String Message = intent.getStringExtra("Message");
                TerminateAlert(Title,Message);
            }
            if(reply.equals("GetAdmin"))
            {
                if (APP.APIResult == APIClass.RESULT_OK) {
                    UpdateListView(APP.APInfoList);
                }
            }
        }
    }
    void SelectListView(AdapterView<?> parent, View view, int position, long id)
    {
        SelectDialog dlg = new SelectDialog(this);
        dlg.setClicklistener(mode->{
            dlg.dismiss();
            Intent intent;
            switch(mode)
            {
                case SelectDialogClicklistener.BTN_GENERAL:
                    intent = new Intent(this,APSettingActivity_general.class);
                    startActivity(intent);
                    break;
                case SelectDialogClicklistener.BTN_IPSET:
                    intent = new Intent(this,APSettingActivity_IP.class);
                    startActivity(intent);
                    break;
            }
        });
        SelectedAPIndex = position;
        isNeedIPSet = true;
        String APName = ((MyApplication.APInfo)List_AP.getAdapter().getItem(position)).apname;
        dlg.show(APName);
    }

    void UpdateListView(MyApplication.APInfo[] apInfos)
    {
        APListAdapter adapter = new APListAdapter(this, apInfos);
        List_AP.setAdapter(adapter);
        Text_AP.setText(format("Total : %3d    ",apInfos.length));
    }
    public void TerminateAlert(String Title,String Messages)
    {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(Title).setMessage(Messages)
                .setPositiveButton("Exit", (dialog, which) -> finish())
                .setOnDismissListener(dialog -> finish()).create().show();
    }
    public void MessageAlert(String Title,String Messages)
    {
        new AlertDialog.Builder(this).setTitle(Title).setMessage(Messages)
                .setPositiveButton("OK", (dialog, which) -> {//OK
                }).create().show();
    }
    public void MessageSnack(String Title,String Messages)
    {
        Snackbar snackbar = Snackbar.make(mainLayout, Title + "\n" + Messages, Snackbar.LENGTH_LONG);

        snackbar.setAction("OK", v->snackbar.dismiss()).show();//OK
    }
}

class APListAdapter extends BaseAdapter
{
    Context mContext = null;
    LayoutInflater layoutInflater = null;
    MyApplication.APInfo[] list;

    public APListAdapter(Context context, MyApplication.APInfo[] data)
    {
        mContext = context;
        list = data;
        layoutInflater = LayoutInflater.from(mContext);
    }

    @Override
    public int getCount() {
        return list.length;
    }

    @Override
    public Object getItem(int position) {
        return list[position];
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        @SuppressLint({"ViewHolder", "InflateParams"})
        View view = layoutInflater.inflate(R.layout.ap_list_item,null);

        TextView time = view.findViewById(R.id.Text1);
        TextView distance = view.findViewById(R.id.Distance);
        ImageView flagImage = view.findViewById(R.id.flagImage);
        ImageView USEImage = view.findViewById(R.id.imageView);

        String Name = list[position].apname;
        final int limit = 20;
        if(Name.length() > limit)Name = Name.substring(0,limit) + "..";
        time.setText(Name);

        if(!("Y".equals(list[position].useyn)))//Notuse일 때 회색/ 이미지 변경(X로)
        {
            //time.setPaintFlags(time.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
            time.setTypeface(Typeface.DEFAULT_BOLD);
            time.setTextColor(Color.parseColor("#7f7f7f"));

            //distance.setPaintFlags(time.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
            distance.setTypeface(Typeface.DEFAULT_BOLD);
            distance.setTextColor(Color.parseColor("#7f7f7f"));

            USEImage.setImageResource(R.drawable.api_list_item_notuse);
        }

        distance.setText(format("%sm",list[position].dist));

        int flag = 0;
        if(list[position].flag.length() == 0)flag = 3;//if no data
        else flag = Integer.parseInt(list[position].flag);

        int flagid = 0;
        if(!(flag>3 || flag<0))
        {
            int[] flagarray = {R.drawable.ap_list_item_flag0,R.drawable.ap_list_item_flag1,R.drawable.ap_list_item_flag2,R.drawable.ap_list_item_flag3};
            flagid = flagarray[flag];
        }
        if(flagid != 0)
        flagImage.setImageResource(flagid);

        return view;
    }
}

File: Android App/MyFirstKotlin/app/src/main/java/com/example/myfirstkotlin/APSettingActivity.kt
package com.example.myfirstkotlin

import android.content.DialogInterface
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Looper
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageButton
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.Lifecycle
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.example.myfirstkotlin.MyApplication.Companion.SelectedidxEquip
import com.example.myfirstkotlin.MyApplication.Companion.colorStateList
import com.google.android.material.bottomnavigation.BottomNavigationView

class APSettingActivity : AppCompatActivity() {
    lateinit var fragment1detail : FragmentAP_1APdetail
    lateinit var fragment2apip : FragmentAP_2APIP
    lateinit var fragment3apgps : FragmentAP_3APGPS

    lateinit var bottomNavigationView : BottomNavigationView
    lateinit var viewpager : ViewPager2
    lateinit var QRButton : ImageButton
    lateinit var backBtn : ImageButton
    lateinit var viewTitle : TextView

    lateinit var toolbar: Toolbar

    lateinit var controller: MyController

    var pagearray = arrayOf(R.id.apDetailItem,R.id.apIPItem,R.id.apGPSItem)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_apsetting)

        if (intent != null) {
            val idxEquipStr = intent.getStringExtra("idxEquip")
            SelectedidxEquip = Integer.parseInt(idxEquipStr)
        }

        controller = MyController(this, Looper.getMainLooper())
        setToolBar()
        viewpager = findViewById(R.id.viewpager)
        bottomNavigationView = findViewById(R.id.bottomNavigationView)
        viewpager.adapter = PagerAdapter(supportFragmentManager, lifecycle,this)
        viewpager.registerOnPageChangeCallback(PageChangeCallback())

        try{
            val item = intent.getIntExtra("focus",0)
            val menuItem = bottomNavigationView.menu.getItem(item)
            viewpager.currentItem = item
            menuItem.isChecked = true
            viewTitle.text = menuItem.title
        }catch(e:IndexOutOfBoundsException){e.printStackTrace()}

        bottomNavigationView.setOnNavigationItemSelectedListener { item -> navigationSelected(item) }

        backBtn = findViewById(R.id.backButton)
        backBtn.setOnClickListener { finish() }

        setColor()
    }
    fun setToolBar()
    {
        toolbar = findViewById(R.id.toolbar) //툴바 설정
        setSupportActionBar(toolbar)
        title = ""
        viewTitle = findViewById(R.id.toolbarText)
        UpdateclientNameTextview()
    }
    fun UpdateclientNameTextview()
    {
        val clientName = MyApplication.getClientName(MyApplication.clientcode ?: "") ?: ""
        val clientnameTextView = findViewById<TextView>(R.id.clientname)
        if(!clientName.isNullOrEmpty()) {
            clientnameTextView.visibility = View.VISIBLE
            clientnameTextView.text = clientName
        }else
            clientnameTextView.visibility = View.GONE
    }
    fun setColor()
    {
        toolbar.setBackgroundColor(Color.parseColor(MyApplication.AppColor))
        toolbar.setTitleTextColor(Color.parseColor(MyApplication.TextColor))

        bottomNavigationView.itemTextColor = colorStateList
        bottomNavigationView.itemRippleColor = colorStateList
        bottomNavigationView.itemIconTintList = colorStateList
    }

    private inner class PagerAdapter(fm: FragmentManager, lc: Lifecycle, m:APSettingActivity): FragmentStateAdapter(fm, lc) {
        val parent = m
        override fun getItemCount() = 3
        override fun createFragment(position: Int): Fragment {
            return when (position) {
                0 -> {fragment1detail = FragmentAP_1APdetail(parent);fragment1detail}
                1 -> {fragment2apip = FragmentAP_2APIP(parent);fragment2apip}
                2 -> {fragment3apgps = FragmentAP_3APGPS(parent);fragment3apgps}
                else -> error("no such position: $position") // TODO : String Conversion
            }
        }
    }
    private fun navigationSelected(item: MenuItem): Boolean {
        item.isChecked = true
        viewTitle.text = item.title//bottomNavigationView.menu.ites.title
        when (item.itemId) {
            R.id.apDetailItem -> viewpager.currentItem = 0
            R.id.apIPItem     -> viewpager.currentItem = 1
            R.id.apGPSItem    -> viewpager.currentItem = 2
            else -> return false
        }
        return true
    }
    private inner class PageChangeCallback: ViewPager2.OnPageChangeCallback() {
        override fun onPageSelected(position: Int) {
            super.onPageSelected(position)
            when (position) {
                0 -> { try{fragment1detail.onFocus()}catch(_:Throwable){}; }
                1 -> { try{fragment2apip.onFocus()}catch(_:Throwable){}; }
                2 -> { try{fragment3apgps.onFocus()}catch(_:Throwable){}; }
                else -> error("no such position: $position") // TODO : String Conversion
            }
            bottomNavigationView.selectedItemId = pagearray[position]
        }
    }
    fun callAPListActivity(focus:Int)
    {
        intent = Intent(this,APSettingListActivity::class.java)
        intent.putExtra("focus",focus)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(intent)
    }

    fun TerminateAlert(Title: String?, Messages: String?) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(Title).setMessage(Messages)
            .setPositiveButton("Exit") { _: DialogInterface?, _: Int -> } // TODO : String Conversion
            .setOnDismissListener { appTerminate() }
            .create().show()
    }

    fun MessageAlert(Title: String?, Messages: String?) {
        AlertDialog.Builder(this).setTitle(Title).setMessage(Messages)
            .setPositiveButton("OK") { _: DialogInterface?, _: Int -> }.create().show()
    } // TODO : String Conversion

    fun appTerminate() {
        finishAffinity()
        System.runFinalization()
        System.exit(0)
    }
}